### DISCUSSION TITLE
Starting an app from a python script

### DISCUSSION THREAD

**Discussion by aranvir** [h2oai/wave#1918] [Q&A] [answered]
#### Is your feature request related to a problem? Please describe

This might be a bit of a corner case, but I'd like to start and stop my app via a `system` service. Since I got my app code and its dependencies installed in a virtual environment, I cannot (easily) use the subprocess solution described below. I tried to import the `cli.run()` function to execute it directly, but due to the click decorators this is not straightforward to use.

#### Describe the solution you'd like

This would be easy to change, by refactoring it into an regular function and a cli wrapper with the click decorators. So, basically how it's done for `uvicorn` as well.

#### Describe alternatives you've considered

What already worked in principle was to start the app as a subprocess with `subprocess.run(['wave', 'run', 'app'])`.
However, that only works from within the environment where wave is installed.

EDIT:
As always, once you ask for help, you find the solution soon after. Using the cli command like `h2o_wave.cli.run(["app", "--no-autostart"])` works

Still, I think the refactoring wouldn't hurt :)

**Comment by mturoci** [comment, marked as answer]
> What already worked in principle was to start the app as a subprocess with subprocess.run(['wave', 'run', 'app']).
However, that only works from within the environment where wave is installed.

The trick is to get the path to your current running python executable via `sys.executable` or manual path, instead of using the global python.

You may specify the path to the `wave` executable, something like `subprocess.run(['venv', 'bin', 'wave', 'run', 'app'])` or run it like `python -m uvicorn --host 0.0.0.0 --port 8000 app.py:main` (also within subprocess) which shall be the equivalent to `wave run`.

> As always, once you ask for help, you find the solution soon after. Using the cli command like h2o_wave.cli.run(["app", "--no-autostart"]) works

Yep, that's a proven fact :). Using `cli.run` is an option as well, but is kinda hacky so I would advise resorting to one of the options above.

**Comment by mturoci** [comment]
Adding a few examples of how we do it

https://github.com/h2oai/wave/blob/db5b17ca912becdadfce2f0faa63500172204b03/e2e/utils.py#L38
https://github.com/h2oai/wave/blob/db5b17ca912becdadfce2f0faa63500172204b03/py/examples/tour.py#L46
https://github.com/h2oai/wave/blob/db5b17ca912becdadfce2f0faa63500172204b03/university/h2o_wave_university/cli.py#L24

**Comment by aranvir** [comment]
Thanks for the quick response and the examples!

I blanked on the option to use `subprocess.run(['venv', 'bin', 'wave', 'run', 'app'])`. I'm gonna use that from now on :)
